src/mission_companion.cpp

```diff
@@ -467,54 +1292,872 @@ bool talk_function::outpost_missions( npc &p, const std::string &id, const std::
     if (cur_key == "Recover Ally from Menial Labor"){
         labor_return(p);
     }
-    if (cur_key == "Assign Ally to Carpentry Work"){
-        individual_mission(p, _("departs to work as a carpenter..."), "_carpenter");
-    }
-    if (cur_key == "Recover Ally from Carpentry Work"){
-        carpenter_return(p);
-    }
-    if (cur_key == "Assign Ally to Forage for Food"){
-        individual_mission(p, _("departs to forage for food..."), "_forage");
+
+    if (cur_key == "Upgrade Camp"){
+        const recipe *making = &recipe_id( bldg ).obj();
+         //Stop upgrade if you don't have materials
+        inventory total_inv = g->u.crafting_inventory();
+        if( making->requirements().can_make_with_inventory( total_inv, 1 ) ){
+            std::vector<std::shared_ptr<npc>> npc_list = companion_list( p, "_faction_upgrade_camp" );
+            int need_food = time_to_food ( time_duration::from_turns( making->time / 100 ) );
+            if( camp_food_supply() < need_food && bldg != "faction_base_camp_1" ){
+                popup( _("You don't have enough food stored to feed your companion.") );
+            } else if (npc_list.empty() ) {
+                npc* comp = individual_mission( p, _("begins to upgrade the camp..."), "_faction_upgrade_camp", false, {},
+                                       making->skill_used.obj().name(), making->difficulty );
+                if( comp != nullptr ){
+                    comp->companion_mission_time_ret = calendar::turn + time_duration::from_turns( making->time/ 100 ) ;
+                    g->u.consume_components_for_craft(making, 1);
+                    g->u.invalidate_crafting_inventory();
+                    camp_food_supply( -need_food );
+                }
+            } else {
+                popup( _("You already have a companion upgrading the camp.") );
+            }
+        } else {
+            popup(_("You don't have the materials for the upgrade.") );
+        }
     }
-    if (cur_key == "Recover Ally from Foraging"){
-        forage_return(p);
+
+    if (cur_key == "Recover Ally from Upgrading" ){
+        upgrade_return( p, omt_pos, "_faction_upgrade_camp" );
     }
 
-    return true;
-}
+    if ( cur_key.find("Recover Ally, [") != std::string::npos && cur_key.find("] Expansion") != std::string::npos ){
```

Comment:
Avoid `string::find` like the plague it is.
Instead store mission type and description separately and only compare the type here, not the entire description.
You can store them as pairs or only store the type and have a function that adds some suffix to the type.